client/mksubvolume.cc

```diff
@@ -259,6 +278,104 @@ find_filesystem(MntTable& mnt_table)
 }
 
 
+string
+get_abs_subvol_path(string subvolume)
+{
+    if(!boost::starts_with(subvolume, "/"))
+	subvolume.insert(0, "/");
+    return subvolume;
+}
+
+
+void
+do_consistency_checks(MntTable& mnt_table, libmnt_fs* fs, libmnt_fs* expected_fs)
+{
+    char *subvol_expected, *subvol_fstab, *subvol_real;
+    const char *dev_expected, *dev_fstab;
+
+    // Set up cache for UUID / LABEL resolution in mnt_table_find_source
+    libmnt_cache* cache = mnt_new_cache();
+    MntTable mtab_table("/");
+    mtab_table.set_cache(cache);
+    mtab_table.parse_mtab();
+
+    if (mnt_fs_get_option(expected_fs, "subvol", &subvol_expected, NULL) != 0)
+	throw runtime_error("mnt_fs_get_option failed");
+
+    // Consistency checks on (partially) existing entries
+    libmnt_fs* fstab_entry = mnt_table.find_target(target, MNT_ITER_FORWARD);
+    libmnt_fs* mounted_entry = mtab_table.find_target(mnt_fs_get_target(expected_fs),
+						      MNT_ITER_BACKWARD);
+    // Map UUID / LABEL to a physical device name
+    dev_expected = mnt_resolve_spec(mnt_fs_get_source(expected_fs), cache);
+    dev_fstab = mnt_resolve_spec(mnt_fs_get_source(fstab_entry), cache);
+    if (dev_expected == NULL)
+	throw runtime_error("parent volume in fstab does not match expected device");
+    if (fstab_entry != NULL && dev_fstab == NULL)
+	throw runtime_error("fstab entry does not map to a real device");
+    if (fstab_entry != NULL && strcmp(dev_fstab, dev_expected) != 0)
+	throw runtime_error("existing fstab entry doesn't match target device");
+    if (fstab_entry != NULL)
+    {
+	if (mnt_fs_get_option(fstab_entry, "subvol", &subvol_fstab, NULL) != 0 ||
+		get_abs_subvol_path(subvol_fstab) != get_abs_subvol_path(subvol_expected))
+	    throw runtime_error("existing fstab entry's subvolume doesn't match");
+    }
+    // Something is mounted there already. Is it the correct device?
+    if (mounted_entry != NULL)
+    {
+	// Find last device in mtab to get the actual mount
+	mounted_entry = mtab_table.find_target(mnt_fs_get_target(expected_fs), MNT_ITER_BACKWARD);
```

Comment:
Isn't the exact statement already used 20 lines above to set ```mount_entry```?

Reply:
Yes, it is - I forgot to remove that after several code refactorings...